File: python-node/raft-node.py
import grpc
import raft_pb2
import raft_pb2_grpc
from concurrent import futures
import random
import time
import threading
import logging
from enum import Enum

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')

class NodeState(Enum):
    FOLLOWER = 1
    CANDIDATE = 2
    LEADER = 3

class RaftNode(raft_pb2_grpc.RaftServiceServicer):
    def __init__(self, node_id, peer_addresses):
        self.id = node_id
        self.peer_addresses = peer_addresses
        self.state = NodeState.FOLLOWER
        self.current_term = 0
        self.voted_for = None
        self.log = []
        self.commit_index = 0
        self.last_applied = 0
        self.election_timeout = random.randint(150, 300) / 1000  # Convert to seconds
        self.last_heartbeat = time.time()
        self.votes_received = set()
        self.leader_id = None
        
        # Start election timer
        self.election_timer = threading.Thread(target=self._run_election_timer)
        self.election_timer.daemon = True
        self.election_timer.start()

    def RequestVote(self, request, context):
        logging.info(f"Process {self.id} receives RPC RequestVote from Process {request.candidateId}")
        
        if request.term < self.current_term:
            return raft_pb2.RequestVoteResponse(term=self.current_term, voteGranted=False)
        
        if request.term > self.current_term:
            self.current_term = request.term
            self.state = NodeState.FOLLOWER
            self.voted_for = None
        
        if (self.voted_for is None or self.voted_for == request.candidateId):
            # Check if candidate's log is at least as up-to-date as receiver's log
            last_log_idx = len(self.log) - 1
            last_log_term = self.log[last_log_idx].term if self.log else 0
            
            if (request.lastLogTerm > last_log_term or 
                (request.lastLogTerm == last_log_term and 
                 request.lastLogIndex >= last_log_idx)):
                self.voted_for = request.candidateId
                self.last_heartbeat = time.time()
                return raft_pb2.RequestVoteResponse(term=self.current_term, voteGranted=True)
        
        return raft_pb2.RequestVoteResponse(term=self.current_term, voteGranted=False)

    def AppendEntries(self, request, context):
        logging.info(f"Process {self.id} receives RPC AppendEntries from Process {request.leaderId}")
        
        if request.term < self.current_term:
            return raft_pb2.AppendEntriesResponse(term=self.current_term, success=False)
        
        self.last_heartbeat = time.time()
        self.leader_id = request.leaderId
        
        if request.term > self.current_term:
            self.current_term = request.term
            self.state = NodeState.FOLLOWER
            self.voted_for = None
        
        # Reset election timeout since we received valid AppendEntries
        self.last_heartbeat = time.time()
        
        return raft_pb2.AppendEntriesResponse(term=self.current_term, success=True)

    def _run_election_timer(self):
        while True:
            time.sleep(0.05)  # Check every 50ms
            
            if self.state == NodeState.LEADER:
                self._send_heartbeat()
                continue
                
            time_since_last_heartbeat = time.time() - self.last_heartbeat
            
            if time_since_last_heartbeat > self.election_timeout:
                self._start_election()

    def _start_election(self):
        self.state = NodeState.CANDIDATE
        self.current_term += 1
        self.voted_for = self.id
        self.votes_received = {self.id}  # Vote for self
        self.last_heartbeat = time.time()
        
        logging.info(f"Process {self.id} starting election for term {self.current_term}")
        
        # Send RequestVote RPCs to all peers
        for peer_id, peer_addr in self.peer_addresses.items():
            if peer_id != self.id:
                try:
                    with grpc.insecure_channel(peer_addr) as channel:
                        stub = raft_pb2_grpc.RaftServiceStub(channel)
                        logging.info(f"Process {self.id} sends RPC RequestVote to Process {peer_id}")
                        response = stub.RequestVote(raft_pb2.RequestVoteRequest(
                            term=self.current_term,
                            candidateId=self.id,
                            lastLogIndex=len(self.log) - 1,
                            lastLogTerm=self.log[-1].term if self.log else 0
                        ))
                        
                        if response.voteGranted:
                            self.votes_received.add(peer_id)
                            
                        if response.term > self.current_term:
                            self.current_term = response.term
                            self.state = NodeState.FOLLOWER
                            self.voted_for = None
                            return
                except Exception as e:
                    logging.error(f"Error sending RequestVote to {peer_id}: {e}")
        
        # Check if we won the election
        if len(self.votes_received) > len(self.peer_addresses) // 2:
            self.state = NodeState.LEADER
            logging.info(f"Process {self.id} became leader for term {self.current_term}")
            self._send_heartbeat()
        else:
            self.state = NodeState.FOLLOWER

    def _send_heartbeat(self):
        for peer_id, peer_addr in self.peer_addresses.items():
            if peer_id != self.id:
                try:
                    with grpc.insecure_channel(peer_addr) as channel:
                        stub = raft_pb2_grpc.RaftServiceStub(channel)
                        logging.info(f"Process {self.id} sends RPC AppendEntries to Process {peer_id}")
                        response = stub.AppendEntries(raft_pb2.AppendEntriesRequest(
                            term=self.current_term,
                            leaderId=self.id,
                            prevLogIndex=len(self.log) - 1,
                            prevLogTerm=self.log[-1].term if self.log else 0,
                            entries=[],
                            leaderCommit=self.commit_index
                        ))
                        
                        if response.term > self.current_term:
                            self.current_term = response.term
                            self.state = NodeState.FOLLOWER
                            self.voted_for = None
                            return
                except Exception as e:
                    logging.error(f"Error sending heartbeat to {peer_id}: {e}")

def serve(node_id, server_address, peer_addresses):
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    raft_node = RaftNode(node_id, peer_addresses)
    raft_pb2_grpc.add_RaftServiceServicer_to_server(raft_node, server)
    server.add_insecure_port(server_address)
    server.start()
    logging.info(f"Process {node_id} started on {server_address}")
    server.wait_for_termination()

if __name__ == '__main__':
    import os
    node_id = int(os.environ.get('NODE_ID', 0))
    server_address = os.environ.get('SERVER_ADDRESS', '[::]:50051')
    peer_addresses = {}  # Will be populated from environment variables
    
    # Parse peer addresses from environment variables
    for env_var in os.environ:
        if env_var.startswith('PEER_'):
            peer_id = int(env_var.split('_')[1])
            peer_addresses[peer_id] = os.environ[env_var]
    
    serve(node_id, server_address, peer_addresses)

File: java-node/src/main/java/RaftNode.java
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

import grpc.*;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.*;
import java.util.logging.Logger;

public class RaftNode extends RaftServiceGrpc.RaftServiceImplBase {
    private static final Logger logger = Logger.getLogger(RaftNode.class.getName());
    
    enum NodeState {
        FOLLOWER,
        CANDIDATE,
        LEADER
    }
    
    private final int nodeId;
    private final Map<Integer, String> peerAddresses;
    private NodeState state;
    private int currentTerm;
    private Integer votedFor;
    private List<LogEntry> log;
    private int commitIndex;
    private int lastApplied;
    private long lastHeartbeat;
    private final long electionTimeout;
    private Set<Integer> votesReceived;
    private Integer leaderId;
    private final ScheduledExecutorService scheduler;
    
    public RaftNode(int nodeId, Map<Integer, String> peerAddresses) {
        this.nodeId = nodeId;
        this.peerAddresses = peerAddresses;
        this.state = NodeState.FOLLOWER;
        this.currentTerm = 0;
        this.votedFor = null;
        this.log = new ArrayList<>();
        this.commitIndex = 0;
        this.lastApplied = 0;
        this.electionTimeout = ThreadLocalRandom.current().nextLong(150, 301);
        this.lastHeartbeat = System.currentTimeMillis();
        this.votesReceived = new HashSet<>();
        this.leaderId = null;
        this.scheduler = Executors.newScheduledThreadPool(1);
        
        // Start election timer
        scheduler.scheduleAtFixedRate(this::checkElectionTimeout, 0, 50, TimeUnit.MILLISECONDS);
    }
    
    @Override
    public void requestVote(RequestVoteRequest request, StreamObserver<RequestVoteResponse> responseObserver) {
        logger.info("Process " + nodeId + " receives RPC RequestVote from Process " + request.getCandidateId());
        
        RequestVoteResponse.Builder response = RequestVoteResponse.newBuilder();
        
        if (request.getTerm() < currentTerm) {
            response.setTerm(currentTerm).setVoteGranted(false);
        } else {
            if (request.getTerm() > currentTerm) {
                currentTerm = request.getTerm();
                state = NodeState.FOLLOWER;
                votedFor = null;
            }
            
            if (votedFor == null || votedFor == request.getCandidateId()) {
                int lastLogIdx = log.size() - 1;
                int lastLogTerm = lastLogIdx >= 0 ? log.get(lastLogIdx).getTerm() : 0;
                
                if (request.getLastLogTerm() > lastLogTerm ||
                    (request.getLastLogTerm() == lastLogTerm &&
                     request.getLastLogIndex() >= lastLogIdx)) {
                    votedFor = request.getCandidateId();
                    lastHeartbeat = System.currentTimeMillis();
                    response.setTerm(currentTerm).setVoteGranted(true);
                } else {
                    response.setTerm(currentTerm).setVoteGranted(false);
                }
            } else {
                response.setTerm(currentTerm).setVoteGranted(false);
            }
        }
        
        responseObserver.onNext(response.build());
        responseObserver.onCompleted();
    }
    
    @Override
    public void appendEntries(AppendEntriesRequest request, StreamObserver<AppendEntriesResponse> responseObserver) {
        logger.info("Process " + nodeId + " receives RPC AppendEntries from Process " + request.getLeaderId());
        
        AppendEntriesResponse.Builder response = AppendEntriesResponse.newBuilder();
        
        if (request.getTerm() < currentTerm) {
            response.setTerm(currentTerm).setSuccess(false);
        } else {
            lastHeartbeat = System.currentTimeMillis();
            leaderId = request.getLeaderId();
            
            if (request.getTerm() > currentTerm) {
                currentTerm = request.getTerm();
                state = NodeState.FOLLOWER;
                votedFor = null;
            }
            
            response.setTerm(currentTerm).setSuccess(true);
        }
        
        responseObserver.onNext(response.build());
        responseObserver.onCompleted();
    }
    
    private void checkElectionTimeout() {
        if (state == NodeState.LEADER) {
            sendHeartbeat();
            return;
        }
        
        long timeSinceLastHeartbeat = System.currentTimeMillis() - lastHeartbeat;
        if (timeSinceLastHeartbeat > electionTimeout) {
            startElection();
        }
    }
    
    private void startElection() {
        state = NodeState.CANDIDATE;
        currentTerm++;
        votedFor = nodeId;
        votesReceived = new HashSet<>(Collections.singleton(nodeId));
        lastHeartbeat = System.currentTimeMillis();
        
        logger.info("Process " + nodeId + " starting election for term " + currentTerm);
        
        for (Map.Entry<Integer, String> peer : peerAddresses.entrySet()) {
            if (peer.getKey() != nodeId) {
                CompletableFuture.runAsync(() -> {
                    try {
                        ManagedChannel channel = ManagedChannelBuilder
                            .forTarget(peer.getValue())
                            .usePlaintext()
                            .build();
                        
                        RaftServiceGrpc.RaftServiceBlockingStub stub = RaftServiceGrpc.newBlockingStub(channel);
                        
                        logger.info("Process " + nodeId + " sends RPC RequestVote to Process " + peer.getKey());
                        
                        RequestVoteRequest request = RequestVoteRequest.newBuilder()
                            .setTerm(currentTerm)
                            .setCandidateId(nodeId)
                            .setLastLogIndex(log.size() - 1)
                            .setLastLogTerm(log.isEmpty() ? 0 : log.get(log.size() - 1).getTerm())
                            .build();
                            
                        RequestVoteResponse response = stub.requestVote(request);
                        
                        if (response.getVoteGranted()) {
                            synchronized (votesReceived) {
                                votesReceived.add(peer.getKey());
                                if (votesReceived.size() > peerAddresses.size() / 2) {
                                    becomeLeader();
                                }
                            }
                        }
                        
                        if (response.getTerm() > currentTerm) {
                            currentTerm = response.getTerm();
                            state = NodeState.FOLLOWER;
                            votedFor = null;
                        }
                        
                        channel.shutdown();
                    } catch (Exception e) {
                        logger.warning("Error sending RequestVote to " + peer.getKey() + ": " + e.getMessage());
                    }
                });
            }
        }
    }
    
    private void becomeLeader() {
        if (state == NodeState.CANDIDATE) {
            state = NodeState.LEADER;
            leaderId = nodeId;
            logger.info("Process " + nodeId + " became leader for term " + currentTerm);
            sendHeartbeat();
        }
    }
    
    private void sendHeartbeat() {
        for (Map.Entry<Integer, String> peer : peerAddresses.entrySet()) {
            if (peer.getKey() != nodeId) {
                CompletableFuture.runAsync(() -> {
                    try {
                        ManagedChannel channel = ManagedChannelBuilder
                            .forTarget(peer.getValue())
                            .usePlaintext()
                            .build();
                        
                        RaftServiceGrpc.RaftServiceBlockingStub stub = RaftServiceGrpc.newBlockingStub(channel);
                        
                        logger.info("Process " + nodeId + " sends RPC AppendEntries to Process " + peer.getKey());
                        
                        AppendEntriesRequest request = AppendEntriesRequest.newBuilder()
                            .setTerm(currentTerm)
                            .setLeaderId(nodeId)
                            .setPrevLogIndex(log.size() - 1)
                            .setPrevLogTerm(log.isEmpty() ? 0 : log.get(log.size() - 1).getTerm())
                            .setLeaderCommit(commitIndex)
                            .build();
                            
                        AppendEntriesResponse response = stub.appendEntries(request);
                        
                        if (response.getTerm() > currentTerm) {
                            currentTerm = response.getTerm();
                            state = NodeState.FOLLOWER;
                            votedFor = null;
                        }
                        
                        channel.shutdown();
                    } catch (Exception e) {
                        logger.warning("Error sending AppendEntries to " + peer.getKey() + ": " + e.getMessage());
                    }
                });
            }
        }
    }
    
    public static void main(String[] args) throws IOException, InterruptedException {
        int nodeId = Integer.parseInt(System.getenv().getOrDefault("NODE_ID", "0"));
        String serverAddress = System.getenv().getOrDefault("SERVER_ADDRESS", "0.0.0.0:50051");
        
        Map<Integer, String> peerAddresses = new HashMap<>();
        
        // Parse peer addresses from environment variables
        System.getenv().forEach((key, value) -> {
            if (key.startsWith("PEER_")) {
                int peerId = Integer.parseInt(key.split("_")[1]);
                peerAddresses.put(peerId, value);
            }
        });
        
        // Create and start the server
        String[] hostPort = serverAddress.split(":");
        Server server = ServerBuilder.forPort(Integer.parseInt(hostPort[1]))
            .addService(new RaftNode(nodeId, peerAddresses))
            .build();
            
        server.start();
        logger.info("Process " + nodeId + " started on " + serverAddress);
        server.awaitTermination();
    }
}
